File: SC2001/Lab/Project 1/mergesort.c
#include <stdio.h>

int keycomparisons = 0;

void printArray(int *array, int n, int m)
{
    for (int i = n; i <= m; i++)
    {
        printf("%d ", array[i]);
    }
    printf("\n");
}

void merge(int *array, int n, int m)
{
    int mid = (n + m) / 2;
    int a = n, b = mid + 1;
    if (m - n <= 0)
        return;
    while (a <= mid && b <= m)
    {
        keycomparisons++;
        if (array[a] <= array[b])
            a++;
        else
        {
            int val = array[b];
            for (int i = b; i > a; i--)
            {
                array[i] = array[i - 1];
            }
            array[a] = val;
            mid++;
            a++;
            b++;
        }
    }
}

void mergeSort(int *array, int start, int end)
{
    int mid = (start + end) / 2;
    if (end - start <= 0)
        return;
    mergeSort(array, start, mid);
    mergeSort(array, mid + 1, end);
    merge(array, start, end);
}

int main()
{
    int array[10] = {6, 1, 15, 3, 22, 3, 88, 10, 90, 4};
    printArray(array, 0, 9);
    mergeSort(array, 0, 9);
    printArray(array, 0, 9);
    printf("%d\n", keycomparisons);
}

File: SC2001/Lab/Project 1/mergesort.py
import os
import datetime
from create_datasets import read_pickled_list


# Definition of a key comparison: one if or if-else statement in merge.

keycomparisons = 0


def merge(arrLeft, arrRight):
    global keycomparisons
    newArr = []
    l = 0
    r = 0
    lenLeft = len(arrLeft)
    lenRight = len(arrRight)
    while l < lenLeft and r < lenRight:
        keycomparisons += 1
        if arrLeft[l] < arrRight[r]:  # Specifically this one.
            newArr.append(arrLeft[l])
            l += 1
        else:
            newArr.append(arrRight[r])
            r += 1
    return newArr + arrLeft[l:] + arrRight[r:]


def mergesort(array):
    length = len(array)
    if length <= 1:
        return array
    mid = length // 2
    arrLeft = array[:mid]
    arrRight = array[mid:]
    arrLeft = mergesort(arrLeft)
    arrRight = mergesort(arrRight)
    newArr = merge(arrLeft, arrRight)
    return newArr


if __name__ == '__main__':
    a = datetime.datetime.now()
    os.chdir("datasets")
    # for i in range(3, 8):
    #     for j in range(1, 6, 4):

    array = read_pickled_list(os.getcwd(), "10^4x1\\1")[0]
    array = mergesort(array)
    # print(array)
    print(keycomparisons)
    b = datetime.datetime.now()
    print(b - a)
